### PR TITLE
Table fillup mode

### PR DESCRIPTION
**Pull request by billy-addepar**
Add table fill up mode for table when total columns width does not match table width. The extra space can go to either the first column or be distributed equally among all columns.

### CODE REVIEW

**Comment by billy-addepar** [comment]
@pzuraq

**Comment by billy-addepar** [comment]
Ready for review

**Comment by cyril-sf** [comment]
Are there tests for the `first_column` and `last_column` modes?

File: addon/components/ember-table.js [hunk 1]
```diff
@@ -252,41 +267,32 @@ export default class EmberTable2 extends Component {
...
-
-        if (delta > 0) {
-          const column = columns.get('lastObject');
-          column.set('width', column.get('width') + delta);
+          const column = columns[i];
+          column.set('width', Math.min(column.get('width') + columnDelta), column.get('minWidth'));
         }
+      } else if (tableFillupMode === TABLE_FILLUP_MODE_LAST_COLUMN) {
+        // Add all delta to last column
+        columns[columns.length - 1].set('width', columns[columns.length - 1].get('width') + delta);
```
Comment (cyril-sf) [inline addon/components/ember-table.js]: I would store `columns[columns.length - 1]` in a variable to make this line easier to read (you're doing it right above)

File: tests/dummy/app/templates/simple.hbs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<div class="demo-container">
+  {{#ember-table
+    rows=rows
+    columns=columns
+    tableFillupMode="equal_column"
+    as |t|
+  }}
+    {{#ember-table-row
+      row=t
+
```
Comment (cyril-sf) [inline tests/dummy/app/templates/simple.hbs]: empty lines are weird

**Comment by billy-addepar** [comment]
Tests added. Comments addressed

File: tests/integration/components/ember-table-test.js [hunk 3]
```diff
@@ -113,10 +114,64 @@ for (const customHeader of customHeaderTests) {
...
+    }
+  });
+
+  test(`Test ${headerTest}column fill up - none mode`, async function(assert) {
+    await setupFullTable(this, {}, { headerComponent: customHeader });
+
+    const headerCount = findAll('.et-thead tr th').length;
+    for (let i = 1; i <= headerCount; i++) {
+      assert.equal(tableHelpers.getHeaderElement(i).offsetWidth, 180,
+        'Table header have same width in proportional fill up mode.');
```
Comment (cyril-sf) [inline tests/integration/components/ember-table-test.js]: this is the `none` mode.

File: tests/integration/components/ember-table-test.js [hunk 4]
```diff
@@ -113,10 +114,64 @@ for (const customHeader of customHeaderTests) {
...
+    }
+  });
+
+  test(`Test ${headerTest}column fill up - none mode`, async function(assert) {
+    await setupFullTable(this, {}, { headerComponent: customHeader });
+
+    const headerCount = findAll('.et-thead tr th').length;
+    for (let i = 1; i <= headerCount; i++) {
+      assert.equal(tableHelpers.getHeaderElement(i).offsetWidth, 180,
```
[the remark this answers is not in the document]
Comment (cyril-sf) [reply tests/integration/components/ember-table-test.js]: Also the sample data doesn't seem to allow to fully test that option, no? (why would they all be the same unless the test data forces that behavior?)

**Comment by billy-addepar** [comment]
Comments addressed

File: addon/components/ember-table.js [hunk 5]
```diff
@@ -252,41 +267,32 @@ export default class EmberTable2 extends Component {
...
         column.set('width', column.get('width') + delta);
-      } else {
-        // Split delta equally among columns.
-        let columnDelta = delta / columns.length;
-        if (columnDelta < 1) {
-          columnDelta = 1;
-        }
+      } else if (tableFillupMode === TABLE_FILLUP_MODE_EQUAL_COLUMN) {
+        // Split delta by their proportion.
+        const columnDelta = delta / columns.length;
```
Comment (cyril-sf) [inline addon/components/ember-table.js]: do we need to do something for the remainder?

Comment (billy-addepar) [reply addon/components/ember-table.js]: This is real number division. You do not have to worry about remainder.

**Comment by billy-addepar** [comment]
Bump

File: addon/components/ember-table.js [hunk 6]
```diff
@@ -54,7 +59,17 @@ export default class EmberTable2 extends Component {
...
+
+  /**
+   * A configuration that controls how columns shrink (or extend) when total column width does not
+   * match table width. Behavior of column modification is as follow:
+   * 1) "none": nothing changed to the column
+   * 2) "equal_column": extra space is distributed equally among all columns
+   * 3) "first_column": extra space is added into the first column.
+   * 4) "last_column": extra space is added into the last column.
+   */
+  @property tableFillupMode = TABLE_FILLUP_MODE_NONE;
```
Comment (pzuraq) [inline addon/components/ember-table.js]: Let's call this `tableResizeMode`, without context I don't think people will understand what "fillup" means here, whereas "resize" gives them a bit more of a hint

**Comment by billy-addepar** [comment]
Name changed.

**Pull request merged**